src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs

```diff
@@ -35,35 +36,22 @@ public class CancellationTokenSource : IDisposable
             ((CancellationTokenSource)obj).NotifyCancellation(throwOnFirstException: false); // skip ThrowIfDisposed() check in Cancel()
         };
 
-        /// <summary>The number of callback partitions to use in a <see cref="CancellationTokenSource"/>. Must be a power of 2.</summary>
-        private static readonly int s_numPartitions = GetPartitionCount();
-        /// <summary><see cref="s_numPartitions"/> - 1, used to quickly mod into <see cref="_callbackPartitions"/>.</summary>
-        private static readonly int s_numPartitionsMask = s_numPartitions - 1;
-
         /// <summary>The current state of the CancellationTokenSource.</summary>
         private volatile int _state;
-        /// <summary>The ID of the thread currently executing the main body of CTS.Cancel()</summary>
-        /// <remarks>
-        /// This helps us to know if a call to ctr.Dispose() is running 'within' a cancellation callback.
-        /// This is updated as we move between the main thread calling cts.Cancel() and any syncContexts
-        /// that are used to actually run the callbacks.
-        /// </remarks>
-        private volatile int _threadIDExecutingCallbacks = -1;
-        /// <summary>Tracks the running callback to assist ctr.Dispose() to wait for the target callback to complete.</summary>
-        private long _executingCallbackId;
-        /// <summary>Partitions of callbacks.  Split into multiple partitions to help with scalability of registering/unregistering; each is protected by its own lock.</summary>
-        private volatile CallbackPartition?[]? _callbackPartitions;
+        /// <summary>Whether this <see cref="CancellationTokenSource"/> has been disposed.</summary>
+        private bool _disposed;
         /// <summary>TimerQueueTimer used by CancelAfter and Timer-related ctors. Used instead of Timer to avoid extra allocations and because the rooted behavior is desired.</summary>
         private volatile TimerQueueTimer? _timer;
         /// <summary><see cref="System.Threading.WaitHandle"/> lazily initialized and returned from <see cref="WaitHandle"/>.</summary>
         private volatile ManualResetEvent? _kernelEvent;
-        /// <summary>Whether this <see cref="CancellationTokenSource"/> has been disposed.</summary>
-        private bool _disposed;
+        /// <summary>Registration state for the source.</summary>
+        /// <remarks>Lazily-initialized, also serving as the lock to protect its contained state.</remarks>
+        private Registrations? _registrations;
 
         // legal values for _state
-        private const int NotCanceledState = 1;
-        private const int NotifyingState = 2;
-        private const int NotifyingCompleteState = 3;
+        private const int NotCanceledState = 0; // default value of _state
+        private const int NotifyingState = 1;
+        private const int NotifyingCompleteState = 2;
```

Comment:
Is it possible this will break any debugger logic? Even if not, I could see this kind of change causing some confusion if someone manually debugging looked up the wrong version of the CTS source code.

Reply:
I'm not aware of any debugging code paying attention to CancellationTokenSource's private _state field and interpreting its value.  In the rare case where we're aware of that (e.g. a few places in Task), we like to annotate it with a comment, and there's no such comment here.